src/assembly/collections.ts

```diff
@@ -0,0 +1,157 @@
+import * as utils from "./utils";
+
+export class CollectionMutationResult {
+  collection!: string;
+  operation!: string;
+  status!: string;
+  key!: string;
+  error!: string;
+}
+
+export class SearchMethodMutationResult {
+  collection!: string;
+  searchMethod!: string;
+  operation!: string;
+  status!: string;
+  error!: string;
+}
+
+export class CollectionSearchResult {
+  collection!: string;
+  searchMethod!: string;
+  status!: string;
+  objects!: CollectionSearchResultObject[];
+  error!: string;
+}
```

Comment:
Lots of duplication here that can be resolved with a base class.  Also we need to consolidate status codes into some enum-like object.

```suggestion
export type CollectionStatus = string;
export namespace CollectionStatus {
  export const Success = "success";
  export const Error = "error";
}

abstract class CollectionResult {
  collection!: string;
  status!: CollectionStatus;
  error!: string;

  get isSuccessful(): bool {
    return this.status == CollectionStatus.Success;
  }
}

export class CollectionMutationResult extends CollectionResult {
  operation!: string;
  key!: string;
}

export class SearchMethodMutationResult extends CollectionResult {
  operation!: string;
  searchMethod!: string;
}

export class CollectionSearchResult extends CollectionResult {
  searchMethod!: string;
  objects!: CollectionSearchResultObject[];
}
```

Reply:
Note that AssemblyScript doesn't yet support pure string enums, so the namespace/type thing is a workaround.  Once declared, you can use them like an enum for assignment or comparison purposes, but the string can still contain any value - it's not restricted to the enum values.  See https://github.com/AssemblyScript/assemblyscript/issues/560